torch/utils/tensorboard/writer.py

```diff
@@ -0,0 +1,731 @@
+                                timestamp):
+        """This adds an entry to the self.scalar_dict datastructure with format
+        {writer_id : [[timestamp, step, value], ...], ...}.
+        """
+        if tag not in self.scalar_dict.keys():
+            self.scalar_dict[tag] = []
+        self.scalar_dict[tag].append(
+            [timestamp, global_step, float(make_np(scalar_value))])
+
+    def _check_caffe2_blob(self, item):
+        """
+        Caffe2 users have the option of passing a string representing the name of
+        a blob in the workspace instead of passing the actual Tensor/array containing
+        the numeric values. Thus, we need to check if we received a string as input
+        instead of an actual Tensor/array, and if so, we need to fetch the Blob
+        from the workspace corresponding to that name. Fetching can be done with the
+        following:
+
+        from caffe2.python import workspace (if not already imported)
+        workspace.FetchBlob(blob_name)
+        workspace.FetchBlobs([blob_name1, blob_name2, ...])
+        """
+        return isinstance(item, six.string_types)
+
+    def get_file_writer(self):
+        """Returns the default FileWriter instance. Recreates it if closed."""
+        if self.all_writers is None or self.file_writer is None:
+            if 'purge_step' in self.kwargs.keys():
+                most_recent_step = self.kwargs.pop('purge_step')
+                self.file_writer = FileWriter(logdir=self.log_dir, **self.kwargs)
+                self.file_writer.add_event(
+                    Event(step=most_recent_step, file_version='brain.Event:2'))
+                self.file_writer.add_event(
+                    Event(step=most_recent_step, session_log=SessionLog(status=SessionLog.START)))
+            else:
+                self.file_writer = FileWriter(logdir=self.log_dir, **self.kwargs)
+            self.all_writers = {self.file_writer.get_logdir(): self.file_writer}
+        return self.file_writer
+
+    def add_scalar(self, tag, scalar_value, global_step=None, walltime=None):
+        """Add scalar data to summary.
+
+        Args:
+            tag (string): Data identifier
+            scalar_value (float or string/blobname): Value to save
+            global_step (int): Global step value to record
+            walltime (float): Optional override default walltime (time.time()) of event
+        """
+        if self._check_caffe2_blob(scalar_value):
+            scalar_value = workspace.FetchBlob(scalar_value)
+        self.get_file_writer().add_summary(
+            scalar(tag, scalar_value), global_step, walltime)
+
+    def add_scalars(self, main_tag, tag_scalar_dict, global_step=None, walltime=None):
+        """Adds many scalar data to summary.
+
+        Note that this function also keeps logged scalars in memory. In extreme case it explodes your RAM.
+
+        Args:
+            main_tag (string): The parent name for the tags
```

Comment:
Huh! Are the tags hierarchical? Where can I read about this in the Tensorboard docs? Is this worth mentioning in the top-level documentation?

Reply:
Yes, tags can be hierarchical. Then the plot will be grouped.
pytorch related doc can be found here:
https://tensorboardx.readthedocs.io/en/latest/tutorial.html#grouping-plots

Reply:
Added a task.